### ISSUE TITLE
[$500] [Wave 6: Tags] Allow for multiple levels of tags on workspace requests

### ISSUE THREAD

**Issue by yuwenmemon** [Expensify/App#33983]
# Part of the [Wave 6: Tags](https://github.com/orgs/Expensify/projects/73) project
Main issue: https://github.com/Expensify/App/issues/32828
Doc section: N/A (Orignal tags doc is here: https://docs.google.com/document/d/1eXGKhdUP98f6RUVYpRzxLi49El9Rizeqq-EYjrF24tU/edit)
Project: https://github.com/orgs/Expensify/projects/73

# Feature Description
We already support multiple levels of tags on policies in Expensify and within our workspaces. Our current code assumes you have a single tag list. We need to update it to juggle the potential for multiple tags. Thankfully, from there, the rest of the framework is already in place. We will do this by:

- [ ] Updating [MoneyRequestConfirmationList](https://github.com/Expensify/App/blob/21d68493954e688d324895ad58ea17ae77166e3c/src/components/MoneyRequestConfirmationList.js) to retrieve all tags in the policy instead of just the first [here](https://github.com/Expensify/App/blob/21d68493954e688d324895ad58ea17ae77166e3c/src/components/MoneyRequestConfirmationList.js#L243-L248). Then, when rendering the [MenuItem](https://github.com/Expensify/App/blob/8aec07c55df3f0fefc6bed76c991183cf6b97f7a/src/components/MenuItemWithTopDescription.tsx) for tag, iteratively render a single MenuItem for each tag [here](https://github.com/Expensify/App/blob/21d68493954e688d324895ad58ea17ae77166e3c/src/components/MoneyRequestConfirmationList.js#L738-L756). The same applies to [MoneyTemporaryForRefactorRequestConfirmationList](https://github.com/Expensify/App/blob/70f3b99c0b5b53d8be84a07f6b0277cbfeda6df4/src/components/MoneyTemporaryForRefactorRequestConfirmationList.js).
- [ ] Do the same thing in [MoneyRequestView](https://github.com/Expensify/App/blob/8c19e21abc950c01a3038a1e8787fa66ae5a6679/src/components/ReportActionItem/MoneyRequestView.js).
- [ ] Adjusting the routes for tags to incorporate the tag index ([discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1703101019479799?thread_ts=1703100843.219419&cid=C01GTK53T8Q)). This means adjusting the current routes for editing/selecting tags in the following way:
    - `${iouType}/new/tag/${reportID}` -> `${iouType}/new/tag/0/${reportID}`
    - `create/${iouType}/tag/${transactionID}/${reportID}/` -> `create/${iouType}/tag/0/${transactionID}/${reportID}/`
    - `r/:threadReportID/edit/:field` -> `r/:threadReportID/edit/:field/0`
- [ ] Then for the tag we send to the API, we create a string of all tags, by order of index, delimited by colons, like so `tag1:tag2:tag3`. The Expensify API already understands this format.
- [ ] For displaying the tag, for instance when we have the `tag` key from [`getTransactionDetails`](https://github.com/Expensify/App/blob/2bb6d00bd6f56273fac83f8e7915751de4f23e62/src/libs/ReportUtils.ts#L1816-L1841), we should split the tag by the colon delimiter and display each in it's corresponding index. 
- [ ] We'll need to create a method that translates a tag index to a tag name. This would correspond to the order a tag appears in the `policyTags_` OnyxKey:
    ![Screenshot 2024-01-04 at 12 06 11 PM](https://github.com/Expensify/App/assets/4741899/7a228340-9ec6-4a62-ae58-e93109078b5f)
- [ ] For any "Modified Expense" message, anytime we change a single tag level, we'll just display the entire colon-delimited tag, like we do on OldDot:
    - <img width="938" alt="Screenshot 2024-01-04 at 2 27 49 PM" src="https://github.com/Expensify/App/assets/4741899/1b860ef4-5e2e-46dc-a85b-222a2b58e62e">

# Manual Test Steps
1. In OldDot, create a policy with multiple tags. One can do this by following the instructions [here](https://community.expensify.com/discussion/5756/how-to-set-up-and-manage-multi-level-tagging#:~:text=You%20can%20enable%20multi%2Dlevel,of%20tags%3F%22%20to%20enabled.), and using the following csv with multiple tag levels:
[independentTags.csv](https://github.com/Expensify/App/files/13834979/independentTags.csv)
2. Create a new workspace request
3. Make sure you see a row rendered for each tag in the workspace
4. Make sure that selection works for each row, and that the appropriate list of tags show for the appropriate level.
5. Save the request, make sure that the tags save properly on the expense
6. Open the recently saved request, make sure that the saved tag selections show up in the correct rows.
7. Edit the request, again make sure that the selections work the same as when you created the expense, and save properly. 

cc @amyevans @rezkiy37 @waterim @puneetlath 

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0131596ecedabdad1c</li>
        <li>Upwork Job ID: 1744420931750514688</li>
        <li>Last Price Increase: 2024-01-08</li>
    </ul>
</details>

**Comment by rezkiy37** [comment]
Hi, I’m Michael (Mykhailo) from Callstack and I would like to work on this issue.

**Comment by yuwenmemon** [comment]
Thanks @rezkiy37! Let me know if you have any questions and if there's any updates you'd like to see on the back-end that we overlooked.

**Comment by rezkiy37** [comment]
I've started to investigate the feature. I will let you know if any questions appear, thank you!

**Comment by rezkiy37** [comment]
Actively working on the issue.

**Comment by rezkiy37** [comment]
@yuwenmemon, what do you think about storing tags in transaction drafts?

<img width="1035" alt="Screenshot 2024-01-11 at 16 37 40" src="https://github.com/Expensify/App/assets/57314631/09b4d827-e212-4f49-bd3a-78dc168c1780">

Also, I consider a transformation from tags object to array. I am going to use the index based on position inside the array.

<img width="571" alt="Screenshot 2024-01-11 at 16 39 54" src="https://github.com/Expensify/App/assets/57314631/c98b57e3-b85f-4506-bdc1-5eae2136ea10">

**Comment by rezkiy37** [comment]
One additional question: how we should handle a case with the indexes, when a new workspace has been added or removed?

**Comment by rezkiy37** [comment]
Waiting on answers for questions above.

**Comment by yuwenmemon** [comment]
> what do you think about storing tags in transaction drafts?

@rezkiy37 I think it would be best to store the tag as a single string delimited by colons, aka `Accounting:North America:Project 7` - as that's what the API is expecting and how the tags are stored in the db:
```
         created = 2023-09-26
   transactionID = 3622990993903110271
          cardID = 2
        reportID = 341010080006119
       receiptID = 
          amount = -2000
        merchant = 3434
             mcc = 0
           state = -3
         comment = {"comment":""}
             tag = Ops:Asia:Project 2
      externalID = 
            memo = 
        inserted = 2023-09-26 05:17:11
        currency = USD
 modifiedCreated = 
  modifiedAmount = 
modifiedMerchant = 
     modifiedMCC = 
        category = Car
 transactionHash = hash2C5B23D78998401AB6C0301F1FBE870D5C55AE4C
modifiedCurrency = 
```

Furthermore, looking at your screenshot I don't think it's necessary to show the GL for now.

**Comment by yuwenmemon** [comment]
> Also, I consider a transformation from tags object to array. I am going to use the index based on position inside the array.

This makes sense to me, yes.

**Comment by yuwenmemon** [comment]
> how we should handle a case with the indexes, when a new workspace has been added or removed?

Hmmm... I'm not quite sure what you mean, sorry. Do you have an example of what you're talking about?

**Comment by rezkiy37** [comment]
> > how we should handle a case with the indexes, when a new workspace has been added or removed?
> 
> Hmmm... I'm not quite sure what you mean, sorry. Do you have an example of what you're talking about?

I mean for example:
1. I am in 5 workspaces.
2. I leave the 3rd workspace.
3. Indexes change.
4. How to handle this case?

**Comment by yuwenmemon** [comment]
Hmmm... how would leaving a workspace affect the indexes of your tags? If you left a workspace we'd remove the tags for that workspace via an Onyx update sent from the server.

**Comment by yuwenmemon** [comment]
If you're asking what should happen when the indexes of the tags within a policy change, here's what we do in OldDot:
Originally I have the order 1. Department 2. Region 3. Project
<img width="568" alt="Screenshot 2024-01-17 at 12 34 07 PM" src="https://github.com/Expensify/App/assets/4741899/5c575c14-c462-4c3d-8f17-9652f976635f">

I change the order to be 1. Region 2. Department 3. Project
<img width="897" alt="Screenshot 2024-01-17 at 12 33 01 PM" src="https://github.com/Expensify/App/assets/4741899/f7178f27-ba6c-4d5f-a7d2-55ec3efb7162">

Since the tag no longer lines up with the order, a violation is shown (we don't need to worry about that for now) and we just show the entire tag string in the space for the first tag.

**Comment by rezkiy37** [comment]
Yeah, makes sense. It is clear now. Thank you!

**Comment by rezkiy37** [comment]
Worked on another task. Going to continue with this one tomorrow.

**Comment by rezkiy37** [comment]
Actively working on the issue. Going to connect the multiple levels of tags with API for creating an expense report.

**Comment by yuwenmemon** [comment]
Thanks @rezkiy37 - do let me know if you need anything from the API that you think is missing.

**Comment by rezkiy37** [comment]
Implementing logic to process a string like tags for rendering and business + API layer.

**Comment by rezkiy37** [comment]
Have a good progress on the issue. Going to adapt recently used tags for the multiple levels feature tomorrow.

https://github.com/Expensify/App/assets/57314631/432a5cfb-1a90-4b1b-9620-106762c1c4c7

**Comment by yuwenmemon** [comment]
Looks awesome @rezkiy37! My only note would be that it looks like we're presenting the GL Codes for the tags as selectable. Technically those are supposed to be linked to the tags themselves and not selectable on their own. Perhaps you're importing them as tag levels themselves though in the video - in that case all good - but just confusing for an expense report accounting veteran like me 😅

**Comment by rezkiy37** [comment]
@yuwenmemon, I just imported those tags and they have "GL" in their titles. So, no tech. features behind it, just naming.

<img width="1449" alt="Screenshot 2024-01-24 at 14 51 22" src="https://github.com/Expensify/App/assets/57314631/b716a4f3-a574-42ba-a5e2-5bfeafb26abb">

**Comment by rezkiy37** [comment]
Integrated editing. Also, applied improvements for the entire flow. Going to test with one-level of tags tomorrow.

**Comment by yuwenmemon** [comment]
> @yuwenmemon, I just imported those tags and they have "GL" in their titles. So, no tech. features behind that, just naming.

Haha okay cool :+1: 😄

**Comment by rezkiy37** [comment]
Worked on another issue today. Will continue on this one tomorrow.

**Comment by rezkiy37** [comment]
Looks like it works good with single level of tags. Implementing violations generation for multiple levels of tags.

**Comment by rezkiy37** [comment]
@yuwenmemon, I've [tried to integrate](https://github.com/Expensify/App/pull/34983/commits/f8e2512a77831ea453215418cfe5717385013d22) violations for tags. Basically, I extended to link tag list name with a tag violation. So, can I ask you to review the commit to understand that it looks properly. 
Also, I see that the feature is under a beta - `violations`, so can I ask you to provide me (mykhailo.kravchenko@callstack.com) the beta?


https://github.com/Expensify/App/assets/57314631/a6f011aa-ec70-4489-8755-4a1449d8932b

**Comment by rezkiy37** [comment]
Integrated the new flow for split bills. Prettified code. Also, integrated logic for modified expense messages. Going to prepare a draft PR tomorrow.

**Comment by rezkiy37** [comment]
Updated tests for violations - https://github.com/Expensify/App/pull/34983/commits/4619e03d4b8d692dba0deb661467eb5a0df7c38e.

**Comment by yuwenmemon** [comment]
@rezkiy37 That looks good to me! Tapping in @cead22 for a second set of eyes.

**Comment by yuwenmemon** [comment]
Added you to the beta!

**Comment by rezkiy37** [comment]
> Added you to the beta!

Still don't have the beta. mykhailo.kravchenko@callstack.com.

<img width="336" alt="Screenshot 2024-02-01 at 11 34 04" src="https://github.com/Expensify/App/assets/57314631/1da3eaab-b02f-455d-8074-0b372ffc9e41">

**Comment by rezkiy37** [comment]
Also, faced a "feature" that violation messages are overlay a tag row. However, looks like it is an expected behaviour. I compared with the `main`.

<img width="1722" alt="Screenshot 2024-02-01 at 11 50 01" src="https://github.com/Expensify/App/assets/57314631/c5a85b6b-a84c-465c-bad1-00a232d95c42">
<img width="1728" alt="Screenshot 2024-02-01 at 11 51 40" src="https://github.com/Expensify/App/assets/57314631/77453e82-e47a-411f-86e2-081947578747">

**Comment by yuwenmemon** [comment]
@rezkiy37 yeah I think that's out of the scope of this PR.

**Comment by yuwenmemon** [comment]
As for the violations beta I see that you've been added - can you confirm the email is right there?
<img width="1469" alt="Screenshot 2024-02-01 at 11 46 50 AM" src="https://github.com/Expensify/App/assets/4741899/c29dd6cc-e5f0-45ca-8545-ce78112cceae">

Perhaps log in and out of the app? Or clear the local Onyx cache entirely?

**Comment by rezkiy37** [comment]
I've got the beta. Thank you!

**Comment by yuwenmemon** [comment]
Thanks for letting me know @thesahindia

**Comment by yuwenmemon** [comment]
Reassigning to @allroundexperts

**Comment by allroundexperts** [comment]
I reviewed https://github.com/Expensify/App/pull/34983 and am eligible for payment I think. Can someone please check?

**Comment by yuwenmemon** [comment]
@Christinadobrzyn Can we pay @allroundexperts for his review?

**Comment by Christinadobrzyn** [comment]
Yes! Payouts due:

- [ ] Contributor+: $500 @allroundexperts (Upwork offer - https://www.upwork.com/nx/wm/offer/102413570)

Upwork job is [here](https://www.upwork.com/jobs/~01796e081ffdf88014).

**Comment by yuwenmemon** [comment]
Thank you!!

**Comment by allroundexperts** [comment]
Hi @Christinadobrzyn!
I get paid through ND. Can you please write a payment summary instead? Thanks!

**Comment by Christinadobrzyn** [comment]
hi! @allroundexperts Thanks for letting me know! The [payment summary](https://github.com/Expensify/App/issues/33983#issuecomment-2123843785) has been updated to be paid through NewDot. Let me know if you need anything else.

**Comment by Christinadobrzyn** [comment]
I think we can close this? @allroundexperts does that sound good?

**Comment by allroundexperts** [comment]
Yep. Good to close!

**Comment by JmillsExpensify** [comment]
$500 approved for @allroundexperts
